<a href="https://colab.research.google.com/github/adhang/data-science-digitalskola/blob/update/08.%20Advanced%20Pandas%20Dataframe/Learn%20-%20Advanced%20Pandas%20Dataframe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Dataframe
Author: Adhang Muntaha Muhammad

[![LinkedIn](https://img.shields.io/badge/linkedin-0077B5?style=for-the-badge&logo=linkedin&logoColor=white&link=https://www.linkedin.com/in/adhangmuntaha/)](https://www.linkedin.com/in/adhangmuntaha/)
[![GitHub](https://img.shields.io/badge/github-121011?style=for-the-badge&logo=github&logoColor=white&link=https://github.com/adhang)](https://github.com/adhang)
[![Kaggle](https://img.shields.io/badge/kaggle-20BEFF?style=for-the-badge&logo=kaggle&logoColor=white&link=https://www.kaggle.com/adhang)](https://www.kaggle.com/adhang)
[![Tableau](https://img.shields.io/badge/tableau-E97627?style=for-the-badge&logo=tableau&logoColor=white&link=https://public.tableau.com/app/profile/adhang)](https://public.tableau.com/app/profile/adhang)
___
**Contents**
- Indexing Dataframe
- Dropping Columns
- Joining Dataframes
- Contatenating Dataframes
- Appending Dataframes
- Pivot Table
- Melting Dataframes
- Lambda Function on Dataframes

# Importing Libraries

In [2]:
import pandas as pd

# Reading Dataset
For this notebook, I will use my mentor's dataset from GitHub. You can check his works [here](https://github.com/densaiko).

In [12]:
file_path = 'https://raw.githubusercontent.com/densaiko/data_science_learning/main/dataset/insurance.csv'

data = pd.read_csv(file_path)
data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

# Indexing Dataframe

## Default Index
An index is used to identify a row/ record. The default index starts from 0 to n (determined by the total rows).

For example, let's see the dataset size using `shape`.

In [4]:
data.shape

(1338, 7)

As we can see, there are 1338 rows and 7 columns. Let's see the first 5 rows.

In [5]:
data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

From this output, we know that the index is started from 0 to 4. How about the last index?

In [6]:
data.tail()

age     sex    bmi  children smoker     region     charges
1333   50    male  30.97         3     no  northwest  10600.5483
1334   18  female  31.92         0     no  northeast   2205.9808
1335   18  female  36.85         0     no  southeast   1629.8335
1336   21  female  25.80         0     no  southwest   2007.9450
1337   61  female  29.07         0    yes  northwest  29141.3603

Now, we know the last 5 rows. The last row has 1337 as the index. So, an index ranging from 0 to 1337 means it has 1338 rows.

## Column as Index
Can we change the index using a column? Yes, absolutely. We can use `set_index()` method to do this.

Let's say we will use the `sex` column as the index.

In [20]:
data.set_index('sex')

age     bmi  children smoker     region      charges
sex                                                         
female   19  27.900         0    yes  southwest  16884.92400
male     18  33.770         1     no  southeast   1725.55230
male     28  33.000         3     no  southeast   4449.46200
male     33  22.705         0     no  northwest  21984.47061
male     32  28.880         0     no  northwest   3866.85520
...     ...     ...       ...    ...        ...          ...
male     50  30.970         3     no  northwest  10600.54830
female   18  31.920         0     no  northeast   2205.98080
female   18  36.850         0     no  southeast   1629.83350
female   21  25.800         0     no  southwest   2007.94500
female   61  29.070         0    yes  northwest  29141.36030

[1338 rows x 6 columns]

Note, it's just for displaying the result. If you want to save it, use `inplace=True` or assign it to a variable (whether it's a new variable or the same variable). Like this:

```
# assign to the original dataframe
data.set_index('sex', inplace=True)

# assign to a new variable
new_data = data.set_index('sex')

# assign to the same variable
data = data.set_index('sex')
```



## Multiple Index
We can set multiple columns as the index. It will create a multi-index.

In [18]:
data.set_index(['sex','smoker'])

age     bmi  children     region      charges
sex    smoker                                               
female yes      19  27.900         0  southwest  16884.92400
male   no       18  33.770         1  southeast   1725.55230
       no       28  33.000         3  southeast   4449.46200
       no       33  22.705         0  northwest  21984.47061
       no       32  28.880         0  northwest   3866.85520
...            ...     ...       ...        ...          ...
       no       50  30.970         3  northwest  10600.54830
female no       18  31.920         0  northeast   2205.98080
       no       18  36.850         0  southeast   1629.83350
       no       21  25.800         0  southwest   2007.94500
       yes      61  29.070         0  northwest  29141.36030

[1338 rows x 5 columns]

Here, the `sex` and `smoker` are set as the index.

## Reset Index
Let's say, our dataframe has an index that confusing (or not ordered). We can reset the index using `reset_index()`.

To demonstrate this, I will create a new dataframe using random sampling. So, if you re-run this notebook, you may get a different result.

In [19]:
data_5 = data.sample(n=5)
data_5

age     sex    bmi  children smoker     region     charges
833    58    male  34.39         0     no  northwest  11743.9341
1065   42  female  25.30         1     no  southwest   7045.4990
976    48    male  40.15         0     no  southeast   7804.1605
721    53    male  36.60         3     no  southwest  11264.5410
785    35  female  27.70         3     no  southwest   6414.1780

This dataframe has an unordered (messy?) index. Let's reset it.

In [21]:
data_5.reset_index()

index  age     sex    bmi  children smoker     region     charges
0    833   58    male  34.39         0     no  northwest  11743.9341
1   1065   42  female  25.30         1     no  southwest   7045.4990
2    976   48    male  40.15         0     no  southeast   7804.1605
3    721   53    male  36.60         3     no  southwest  11264.5410
4    785   35  female  27.70         3     no  southwest   6414.1780

We can see that the index has reset. But wait. There's a new column named `index`. This column contains the previous index. How to reset the index without creating a new column? We can pass the `drop=True` parameter.

In [22]:
data_5.reset_index(drop=True)

age     sex    bmi  children smoker     region     charges
0   58    male  34.39         0     no  northwest  11743.9341
1   42  female  25.30         1     no  southwest   7045.4990
2   48    male  40.15         0     no  southeast   7804.1605
3   53    male  36.60         3     no  southwest  11264.5410
4   35  female  27.70         3     no  southwest   6414.1780

Now, our dataframe contains a nice index, right?

# Dropping Columns and Rows
Sometimes, for some reason, we have to delete or remove some columns or rows from our dataset.

In the previous notebook (Intermediate Pandas Dataframe), I had written some ways to select some columns or rows such as using `loc`, `iloc`, etc. But in some cases, dropping some columns is easier than selecting some columns.

## Dropping Columns
Let's say, we want to select all columns except the `charges` column. To drop a column, don't forget to add `axis=1` since columns are axis 1.

In [24]:
drop_col_1 = data.drop('charges', axis=1)
drop_col_1.head()

age     sex     bmi  children smoker     region
0   19  female  27.900         0    yes  southwest
1   18    male  33.770         1     no  southeast
2   28    male  33.000         3     no  southeast
3   33    male  22.705         0     no  northwest
4   32    male  28.880         0     no  northwest

To drop multiple columns, we can use a list containing the column names.

In [25]:
drop_col_2 = data.drop(['age','sex'], axis=1)
drop_col_2.head()

bmi  children smoker     region      charges
0  27.900         0    yes  southwest  16884.92400
1  33.770         1     no  southeast   1725.55230
2  33.000         3     no  southeast   4449.46200
3  22.705         0     no  northwest  21984.47061
4  28.880         0     no  northwest   3866.85520

## Dropping Rows
To drop a row, we have to specify the index number and set the axis parameter to 0.

For example, we will drop a row that has index `1`.

In [26]:
drop_row = data.drop(1, axis=0)
drop_row.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520
5   31  female  25.740         0     no  southeast   3756.62160

Now, the row that has index `1` has been removed.

What if the index has the same values? Let's say, the `sex` column is set as the index. Then, we remove the rows that have `female` as the index. What will happen?

In [27]:
data_sex = data.set_index('sex')
data_sex.head()

age     bmi  children smoker     region      charges
sex                                                         
female   19  27.900         0    yes  southwest  16884.92400
male     18  33.770         1     no  southeast   1725.55230
male     28  33.000         3     no  southeast   4449.46200
male     33  22.705         0     no  northwest  21984.47061
male     32  28.880         0     no  northwest   3866.85520

In [28]:
data_sex_drop = data_sex.drop('female', axis=0)
data_sex_drop.head()

age     bmi  children smoker     region      charges
sex                                                       
male   18  33.770         1     no  southeast   1725.55230
male   28  33.000         3     no  southeast   4449.46200
male   33  22.705         0     no  northwest  21984.47061
male   32  28.880         0     no  northwest   3866.85520
male   37  29.830         2     no  northeast   6406.41070

As you can see, all the rows that have `female` as the index are removed.

# Joining Dataframes
In my previous notebook (Intermediate Pandas Dataframe), we can combine several tables in terms of columns (combining both tables' columns) using the `merge()` method.
<br><br>
Pandas has another merging/ joining method named `join()`. It has the same function as `merge()`. But, `join()` will only combine the dataframes based on the rows number (index). It's just like row's matching.
<br><br>
For example, I will create 2 new dataframes. The first one contains 10 records using random sampling. And the second one is half of the first dataframe (the first 5 rows).
<br><br>
**Note:** if you re-run this notebook, you may get a different result.

In [36]:
data_left = data.sample(n=10)
data_right = data_left.iloc[:5]

display(data_left)
display(data_right)

age     sex    bmi  children smoker     region      charges
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760
1234   45  female  27.83         2     no  southeast   8515.75870
547    54  female  46.70         2     no  southwest  11538.42100
684    33  female  18.50         1     no  southwest   4766.02200
149    19    male  28.40         1     no  southwest   1842.51900
164    37    male  29.64         0     no  northwest   5028.14660

age     sex    bmi  children smoker     region      charges
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760

As you can see, the second dataframe contains the first 5 rows from the first dataframe. Let's do a join.
<br><br>
We need to specify `lsuffix` (left suffix) and `rsuffix` (right suffix). If the both tables have the same column names, it's used to identify whether the columns are from the left dataframe or the right dataframe.

In [37]:
data_left.join(data_right, lsuffix='_first', rsuffix='_second')

age_first sex_first  ...  region_second  charges_second
1046         43    female  ...      northeast      7325.04820
925          50      male  ...      northeast     25333.33284
403          49      male  ...      northwest     10269.46000
1152         43    female  ...      southeast     40941.28540
1062         59      male  ...      southeast     48970.24760
1234         45    female  ...            NaN             NaN
547          54    female  ...            NaN             NaN
684          33    female  ...            NaN             NaN
149          19      male  ...            NaN             NaN
164          37      male  ...            NaN             NaN

[10 rows x 14 columns]

The first 5 rows contain full records because the data (that have the same index) appears on both dataframes.
<br><br>
The column names contain `_first` or `_second` postfix, except the `index` column. That's because, we are joining 2 tables with `index` as a connector, which means that the columns are considered as the same column. The other columns are considered as different columns, even the column names are the same. There is a reason behind it. Let's see the 2 tables below.

> **order_table**

| **order_id** | **user_id** |  **name**  | **price** |
|:------------:|:-----------:|:----------:|:---------:|
|       1      |      2      |    mango   |    4000   |
|       2      |      1      | strawberry |    5000   |
|       3      |      1      |    lemon   |    4500   |

> **user_table**

| **user_id** | **name** |  **city** |
|:-----------:|:--------:|:---------:|
|      1      |  Adhang  |   Jogja   |
|      2      |  Muntaha |  Jakarta  |
|      3      | Muhammad | Palembang |

Both tables contain a `name` column. What if we merge those tables and assume that the `name` column is the same for both tables? It will be a disaster because the `name` column on `order_table` represents the product names, but the `name` column on `user_table` represents the user names.
<br><br>
To handle this problem, Pandas is assuming that they are different column even the column name are the same. So, Pandas automatically add a postfix to represent it as a different column.

# Concatenating Dataframes
Concatenating is used to combine several tables in terms of rows or columns (combining both tables' rows or columns). The mode is determined by the axis parameter. Axis 0 is combining the rows, and axis 1 is combining the columns.

## Concatenate Columns
Concatenating columns is like `join()` method, we will combine the dataframes' column based on its index.

In [46]:
display(data_left.sort_index())
display(data_right.sort_index())

age     sex    bmi  children smoker     region      charges
149    19    male  28.40         1     no  southwest   1842.51900
164    37    male  29.64         0     no  northwest   5028.14660
403    49    male  32.30         3     no  northwest  10269.46000
547    54  female  46.70         2     no  southwest  11538.42100
684    33  female  18.50         1     no  southwest   4766.02200
925    50    male  32.11         2     no  northeast  25333.33284
1046   43  female  25.08         0     no  northeast   7325.04820
1062   59    male  41.14         1    yes  southeast  48970.24760
1152   43  female  32.56         3    yes  southeast  40941.28540
1234   45  female  27.83         2     no  southeast   8515.75870

age     sex    bmi  children smoker     region      charges
403    49    male  32.30         3     no  northwest  10269.46000
925    50    male  32.11         2     no  northeast  25333.33284
1046   43  female  25.08         0     no  northeast   7325.04820
1062   59    male  41.14         1    yes  southeast  48970.24760
1152   43  female  32.56         3    yes  southeast  40941.28540

In [48]:
pd.concat([data_left, data_right], axis=1)

age     sex    bmi children  ... children smoker     region      charges
149    19    male  28.40        1  ...      NaN    NaN        NaN          NaN
164    37    male  29.64        0  ...      NaN    NaN        NaN          NaN
403    49    male  32.30        3  ...      3.0     no  northwest  10269.46000
547    54  female  46.70        2  ...      NaN    NaN        NaN          NaN
684    33  female  18.50        1  ...      NaN    NaN        NaN          NaN
925    50    male  32.11        2  ...      2.0     no  northeast  25333.33284
1046   43  female  25.08        0  ...      0.0     no  northeast   7325.04820
1062   59    male  41.14        1  ...      1.0    yes  southeast  48970.24760
1152   43  female  32.56        3  ...      3.0    yes  southeast  40941.28540
1234   45  female  27.83        2  ...      NaN    NaN        NaN          NaN

[10 rows x 14 columns]

One thing that makes it slightly different from `join()` is that the dataframe is sorted by its index.

## Concatenate Rows
Concatenating rows is like adding rows from another table.

In [49]:
pd.concat([data_left, data_right], axis=0)

age     sex    bmi  children smoker     region      charges
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760
1234   45  female  27.83         2     no  southeast   8515.75870
547    54  female  46.70         2     no  southwest  11538.42100
684    33  female  18.50         1     no  southwest   4766.02200
149    19    male  28.40         1     no  southwest   1842.51900
164    37    male  29.64         0     no  northwest   5028.14660
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760

What if the dataframes have different columns? Let's say the second dataframe only contains `age` and `sex` columns.

In [50]:
pd.concat([data_left, data_right.loc[:,['age','sex']]], axis=0)

age     sex    bmi  children smoker     region      charges
1046   43  female  25.08       0.0     no  northeast   7325.04820
925    50    male  32.11       2.0     no  northeast  25333.33284
403    49    male  32.30       3.0     no  northwest  10269.46000
1152   43  female  32.56       3.0    yes  southeast  40941.28540
1062   59    male  41.14       1.0    yes  southeast  48970.24760
1234   45  female  27.83       2.0     no  southeast   8515.75870
547    54  female  46.70       2.0     no  southwest  11538.42100
684    33  female  18.50       1.0     no  southwest   4766.02200
149    19    male  28.40       1.0     no  southwest   1842.51900
164    37    male  29.64       0.0     no  northwest   5028.14660
1046   43  female    NaN       NaN    NaN        NaN          NaN
925    50    male    NaN       NaN    NaN        NaN          NaN
403    49    male    NaN       NaN    NaN        NaN          NaN
1152   43  female    NaN       NaN    NaN        NaN          NaN
1062   59    male    NaN       NaN    NaN        NaN          NaN

The other columns that didn't match will have `NaN` values, as you can see on the last 5 rows.

# Appending Dataframes
Appending dataframes is like concatenating on axis 0 (row).

In [51]:
data_left.append(data_right)

age     sex    bmi  children smoker     region      charges
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760
1234   45  female  27.83         2     no  southeast   8515.75870
547    54  female  46.70         2     no  southwest  11538.42100
684    33  female  18.50         1     no  southwest   4766.02200
149    19    male  28.40         1     no  southwest   1842.51900
164    37    male  29.64         0     no  northwest   5028.14660
1046   43  female  25.08         0     no  northeast   7325.04820
925    50    male  32.11         2     no  northeast  25333.33284
403    49    male  32.30         3     no  northwest  10269.46000
1152   43  female  32.56         3    yes  southeast  40941.28540
1062   59    male  41.14         1    yes  southeast  48970.24760

See? It's just the same.

# Merge vs Join vs Concat vs Append
You may notice that there are several methods to combine tables. Here are the of thumb:
- `merge` - used to combine tables' columns based on connector (specific column)
- `join` - used to combine tables' columns based on the index
- `append` - used to combine tables' rows
- `concat`
  - Axis 1 - used to combine tables' columns based on the index, similar to `join`
  - Axis 0 - used to combine tables' rows, similar to `append`

# Pivot Table

In [ ]:
data.set_index(['sex','smoker'])

age     bmi  children     region      charges
sex    smoker                                               
female yes      19  27.900         0  southwest  16884.92400
male   no       18  33.770         1  southeast   1725.55230
       no       28  33.000         3  southeast   4449.46200
       no       33  22.705         0  northwest  21984.47061
       no       32  28.880         0  northwest   3866.85520
...            ...     ...       ...        ...          ...
       no       50  30.970         3  northwest  10600.54830
female no       18  31.920         0  northeast   2205.98080
       no       18  36.850         0  southeast   1629.83350
       no       21  25.800         0  southwest   2007.94500
       yes      61  29.070         0  northwest  29141.36030

[1338 rows x 5 columns]

In [ ]:
data.groupby(['sex','smoker','region']).age.mean().unstack()

region         northeast  northwest  southeast  southwest
sex    smoker                                            
female no      39.840909  39.755556  39.071942  40.099291
       yes     38.724138  38.827586  39.250000  37.047619
male   no      39.216000  38.568182  38.261194  40.277778
       yes     37.868421  39.827586  40.054545  35.567568

In [ ]:
pd.pivot_table(data, values='age', index=['sex','smoker'], columns='region', aggfunc='mean')

region         northeast  northwest  southeast  southwest
sex    smoker                                            
female no      39.840909  39.755556  39.071942  40.099291
       yes     38.724138  38.827586  39.250000  37.047619
male   no      39.216000  38.568182  38.261194  40.277778
       yes     37.868421  39.827586  40.054545  35.567568

In [ ]:
pd.pivot_table(data, values='age', index=['sex','smoker'], columns='region', aggfunc=['mean','max'])

mean                        ...       max                    
region         northeast  northwest  southeast  ... northwest southeast southwest
sex    smoker                                   ...                              
female no      39.840909  39.755556  39.071942  ...        64        64        64
       yes     38.724138  38.827586  39.250000  ...        64        64        64
male   no      39.216000  38.568182  38.261194  ...        64        64        64
       yes     37.868421  39.827586  40.054545  ...        62        64        61

[4 rows x 8 columns]